File: test/language/choice_types/python/UInt64ParamChoiceTest.py
import unittest
import os
import zserio

from testutils import getZserioApi, getApiDir

class UInt64ParamChoiceTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "choice_types.zs").uint64_param_choice

    def testConstructor(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(self.VARIANT_A_SELECTOR, uint64ParamChoice.selector)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, b_=1234)
        self.assertEqual(self.VARIANT_B_SELECTOR, uint64ParamChoice.selector)
        self.assertEqual(1234, uint64ParamChoice.b)

    def testFromReader(self):
        selector = self.VARIANT_B_SELECTOR
        value = 234
        writer = zserio.BitStreamWriter()
        UInt64ParamChoiceTest._writeUInt64ParamChoiceToStream(writer, selector, value)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        uint64ParamChoice = self.api.UInt64ParamChoice.from_reader(reader, selector)
        self.assertEqual(selector, uint64ParamChoice.selector)
        self.assertEqual(value, uint64ParamChoice.b)

    def testEq(self):
        uint64ParamChoice1 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        uint64ParamChoice2 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertTrue(uint64ParamChoice1 == uint64ParamChoice2)

        value = 99
        uint64ParamChoice1.a = value
        self.assertFalse(uint64ParamChoice1 == uint64ParamChoice2)

        uint64ParamChoice2.a = value
        self.assertTrue(uint64ParamChoice1 == uint64ParamChoice2)

        diffValue = value + 1
        uint64ParamChoice2.a = diffValue
        self.assertFalse(uint64ParamChoice1 == uint64ParamChoice2)

    def testHash(self):
        uint64ParamChoice1 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        uint64ParamChoice2 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(hash(uint64ParamChoice1), hash(uint64ParamChoice2))

        value = 99
        uint64ParamChoice1.a = value
        self.assertTrue(hash(uint64ParamChoice1) != hash(uint64ParamChoice2))

        uint64ParamChoice2.a = value
        self.assertEqual(hash(uint64ParamChoice1), hash(uint64ParamChoice2))

        diffValue = value + 1
        uint64ParamChoice2.a = diffValue
        self.assertTrue(hash(uint64ParamChoice1) != hash(uint64ParamChoice2))

    def testGetSelector(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(self.VARIANT_C_SELECTOR, uint64ParamChoice.selector)

    def testGetSetA(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        value = 99
        uint64ParamChoice.a = value
        self.assertEqual(value, uint64ParamChoice.a)

    def testGetSetB(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        value = 234
        uint64ParamChoice.b = value
        self.assertEqual(value, uint64ParamChoice.b)

    def testGetSetC(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        value = 23456
        uint64ParamChoice.c = value
        self.assertEqual(value, uint64ParamChoice.c)

    def testChoiceTag(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_A, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_B, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_C, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.EMPTY_SELECTOR)
        self.assertEqual(uint64ParamChoice.UNDEFINED_CHOICE, uint64ParamChoice.choice_tag)

    def testBitSizeOf(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(8, uint64ParamChoice.bitsizeof())

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(16, uint64ParamChoice.bitsizeof())

    def testInitializeOffsets(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        bitPosition = 1
        self.assertEqual(9, uint64ParamChoice.initialize_offsets(bitPosition))

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(17, uint64ParamChoice.initialize_offsets(bitPosition))

    def testWriteRead(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        byteValue = 99
        uint64ParamChoice.a = byteValue
        writer = zserio.BitStreamWriter()
        uint64ParamChoice.write(writer)
        readUInt64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt64ParamChoice.read(reader)
        self.assertEqual(byteValue, readUInt64ParamChoice.a)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

        shortValue = 234
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, b_=shortValue)
        writer = zserio.BitStreamWriter()
        uint64ParamChoice.write(writer)
        readUInt64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt64ParamChoice.read(reader)
        self.assertEqual(shortValue, readUInt64ParamChoice.b)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

    def testWriteReadFile(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        byteValue = 99
        uint64ParamChoice.a = byteValue
        filenameA = self.BLOB_NAME_BASE + "a.blob"
        zserio.serialize_to_file(uint64ParamChoice, filenameA)

        readUInt64ParamChoice = zserio.deserialize_from_file(self.api.UInt64ParamChoice, filenameA,
                                                             self.VARIANT_A_SELECTOR)
        self.assertEqual(byteValue, readUInt64ParamChoice.a)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

        shortValue = 234
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, b_=shortValue)
        filenameB = self.BLOB_NAME_BASE + "b.blob"
        zserio.serialize_to_file(uint64ParamChoice, filenameB)

        readUInt64ParamChoice = zserio.deserialize_from_file(self.api.UInt64ParamChoice, filenameB,
                                                             self.VARIANT_B_SELECTOR)
        self.assertEqual(shortValue, readUInt64ParamChoice.b)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

    @staticmethod
    def _writeUInt64ParamChoiceToStream(writer, selector, value):
        if selector == 1:
            writer.write_signed_bits(value, 8)
        elif selector in (2, 3, 4):
            writer.write_signed_bits(value, 16)
        elif selector in (5, 6):
            pass
        else:
            writer.write_signed_bits(value, 32)

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "uint64_param_choice_")
    VARIANT_A_SELECTOR = 1
    VARIANT_B_SELECTOR = 2
    VARIANT_C_SELECTOR = 7
    EMPTY_SELECTOR = 5

File: test/language/choice_types/cpp/UInt64ParamChoiceTest.cpp
#include "gtest/gtest.h"

#include "zserio/SerializeUtil.h"

#include "choice_types/uint64_param_choice/UInt64ParamChoice.h"

namespace choice_types
{
namespace uint64_param_choice
{

class UInt64ParamChoiceTest : public ::testing::Test
{
protected:
    void writeUInt64ParamChoiceToByteArray(zserio::BitStreamWriter& writer, uint64_t selector,
            int32_t value)
    {
        switch (selector)
        {
        case 1:
            writer.writeSignedBits(value, 8);
            break;

        case 2:
        case 3:
        case 4:
            writer.writeSignedBits(value, 16);
            break;

        case 5:
        case 6:
            break;

        default:
            writer.writeSignedBits(value, 32);
            break;
        }
    }

    static const std::string BLOB_NAME_BASE;
    static const uint64_t VARIANT_A_SELECTOR = 1;
    static const uint64_t VARIANT_B_SELECTOR = 2;
    static const uint64_t VARIANT_C_SELECTOR = 7;
    static const uint64_t EMPTY_SELECTOR = 5;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::string UInt64ParamChoiceTest::BLOB_NAME_BASE = "language/choice_types/uint64_param_choice_";

TEST_F(UInt64ParamChoiceTest, emptyConstructor)
{
    {
        UInt64ParamChoice uint64ParamChoice;
        ASSERT_THROW(uint64ParamChoice.getSelector(), zserio::CppRuntimeException);
    }
    {
        UInt64ParamChoice uint64ParamChoice = {};
        ASSERT_THROW(uint64ParamChoice.getSelector(), zserio::CppRuntimeException);
    }
}

TEST_F(UInt64ParamChoiceTest, bitStreamReaderConstructor)
{
    const uint64_t selector = VARIANT_A_SELECTOR;
    const int8_t value = 99;
    zserio::BitStreamWriter writer(bitBuffer);
    writeUInt64ParamChoiceToByteArray(writer, selector, value);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    UInt64ParamChoice uint64ParamChoice(reader, selector);
    ASSERT_EQ(selector, uint64ParamChoice.getSelector());
    ASSERT_EQ(value, uint64ParamChoice.getA());
}

TEST_F(UInt64ParamChoiceTest, copyConstructor)
{
    const uint64_t selector = VARIANT_A_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    const int8_t value = 99;
    uint64ParamChoice.setA(value);

    const UInt64ParamChoice uint64ParamChoiceCopy(uint64ParamChoice);
    ASSERT_EQ(selector, uint64ParamChoiceCopy.getSelector());
    ASSERT_EQ(value, uint64ParamChoiceCopy.getA());
}

TEST_F(UInt64ParamChoiceTest, assignmentOperator)
{
    const uint64_t selector = VARIANT_B_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    const int16_t value = 234;
    uint64ParamChoice.setB(value);

    UInt64ParamChoice uint64ParamChoiceCopy;
    uint64ParamChoiceCopy = uint64ParamChoice;
    ASSERT_EQ(selector, uint64ParamChoiceCopy.getSelector());
    ASSERT_EQ(value, uint64ParamChoiceCopy.getB());
}

TEST_F(UInt64ParamChoiceTest, moveConstructor)
{
    const uint64_t selector = VARIANT_A_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    const int8_t value = 99;
    uint64ParamChoice.setA(value);

    const UInt64ParamChoice uint64ParamChoiceMoved(std::move(uint64ParamChoice));
    ASSERT_EQ(selector, uint64ParamChoiceMoved.getSelector());
    ASSERT_EQ(value, uint64ParamChoiceMoved.getA());
}

TEST_F(UInt64ParamChoiceTest, moveAssignmentOperator)
{
    const uint64_t selector = VARIANT_B_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    const int16_t value = 234;
    uint64ParamChoice.setB(value);

    UInt64ParamChoice uint64ParamChoiceMoved;
    uint64ParamChoiceMoved = std::move(uint64ParamChoice);
    ASSERT_EQ(selector, uint64ParamChoiceMoved.getSelector());
    ASSERT_EQ(value, uint64ParamChoiceMoved.getB());
}

TEST_F(UInt64ParamChoiceTest, propagateAllocatorCopyConstructor)
{
    const uint64_t selector = VARIANT_A_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    const int8_t value = 99;
    uint64ParamChoice.setA(value);

    const UInt64ParamChoice uint64ParamChoiceCopy(zserio::PropagateAllocator, uint64ParamChoice,
            UInt64ParamChoice::allocator_type());
    ASSERT_EQ(selector, uint64ParamChoiceCopy.getSelector());
    ASSERT_EQ(value, uint64ParamChoiceCopy.getA());
}

TEST_F(UInt64ParamChoiceTest, initialize)
{
    const uint64_t selector = VARIANT_B_SELECTOR;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    ASSERT_EQ(selector, uint64ParamChoice.getSelector());
}

TEST_F(UInt64ParamChoiceTest, getSelector)
{
    const uint64_t selector = 1;
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(selector);
    ASSERT_EQ(selector, uint64ParamChoice.getSelector());
}

TEST_F(UInt64ParamChoiceTest, getSetA)
{
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(VARIANT_A_SELECTOR);
    const int8_t value = 99;
    uint64ParamChoice.setA(value);
    ASSERT_EQ(value, uint64ParamChoice.getA());
}

TEST_F(UInt64ParamChoiceTest, getSetB)
{
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(VARIANT_B_SELECTOR);
    const int16_t value = 234;
    uint64ParamChoice.setB(value);
    ASSERT_EQ(value, uint64ParamChoice.getB());
}

TEST_F(UInt64ParamChoiceTest, getSetC)
{
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(VARIANT_C_SELECTOR);
    const int32_t value = 23456;
    uint64ParamChoice.setC(value);
    ASSERT_EQ(value, uint64ParamChoice.getC());
}

TEST_F(UInt64ParamChoiceTest, choiceTag)
{
    UInt64ParamChoice uint64ParamChoice;
    uint64ParamChoice.initialize(VARIANT_A_SELECTOR);
    ASSERT_EQ(UInt64ParamChoice::CHOICE_a, uint64ParamChoice.choiceTag());

    uint64ParamChoice.initialize(VARIANT_B_SELECTOR);
    ASSERT_EQ(UInt64ParamChoice::CHOICE_b, uint64ParamChoice.choiceTag());

    uint64ParamChoice.initialize(VARIANT_C_SELECTOR);
    ASSERT_EQ(UInt64ParamChoice::CHOICE_c, uint64ParamChoice.choiceTag());

    uint64ParamChoice.initialize(EMPTY_SELECTOR);
    ASSERT_EQ(UInt64ParamChoice::UNDEFINED_CHOICE, uint64ParamChoice.choiceTag());
}

TEST_F(UInt64ParamChoiceTest, bitSizeOf)
{
    UInt64ParamChoice uint64ParamChoiceB;
    uint64ParamChoiceB.initialize(VARIANT_A_SELECTOR);
    ASSERT_EQ(8, uint64ParamChoiceB.bitSizeOf());

    UInt64ParamChoice uint64ParamChoiceG;
    uint64ParamChoiceG.initialize(VARIANT_B_SELECTOR);
    ASSERT_EQ(16, uint64ParamChoiceG.bitSizeOf());
}

TEST_F(UInt64ParamChoiceTest, initializeOffsets)
{
    UInt64ParamChoice uint64ParamChoiceB;
    uint64ParamChoiceB.initialize(VARIANT_A_SELECTOR);
    const size_t bitPosition = 1;
    ASSERT_EQ(9, uint64ParamChoiceB.initializeOffsets(bitPosition));

    UInt64ParamChoice uint64ParamChoiceG;
    uint64ParamChoiceG.initialize(VARIANT_B_SELECTOR);
    ASSERT_EQ(17, uint64ParamChoiceG.initializeOffsets(bitPosition));
}

TEST_F(UInt64ParamChoiceTest, operatorEquality)
{
    UInt64ParamChoice uint64ParamChoice1;
    uint64ParamChoice1.initialize(VARIANT_A_SELECTOR);
    UInt64ParamChoice uint64ParamChoice2;
    uint64ParamChoice2.initialize(VARIANT_A_SELECTOR);
    ASSERT_TRUE(uint64ParamChoice1 == uint64ParamChoice2);

    const int8_t value = 99;
    uint64ParamChoice1.setA(value);
    ASSERT_FALSE(uint64ParamChoice1 == uint64ParamChoice2);

    uint64ParamChoice2.setA(value);
    ASSERT_TRUE(uint64ParamChoice1 == uint64ParamChoice2);

    const int8_t diffValue = value + 1;
    uint64ParamChoice2.setA(diffValue);
    ASSERT_FALSE(uint64ParamChoice1 == uint64ParamChoice2);
}

TEST_F(UInt64ParamChoiceTest, operatorLessThan)
{
    UInt64ParamChoice uint64ParamChoice1;
    uint64ParamChoice1.initialize(VARIANT_A_SELECTOR);
    UInt64ParamChoice uint64ParamChoice2;
    uint64ParamChoice2.initialize(VARIANT_B_SELECTOR);
    ASSERT_TRUE(uint64ParamChoice1 < uint64ParamChoice2);
    ASSERT_FALSE(uint64ParamChoice2 < uint64ParamChoice1);

    uint64ParamChoice2.initialize(VARIANT_A_SELECTOR);
    ASSERT_FALSE(uint64ParamChoice1 < uint64ParamChoice2);
    ASSERT_FALSE(uint64ParamChoice2 < uint64ParamChoice1);

    const int8_t value = 99;
    uint64ParamChoice1.setA(value);
    ASSERT_FALSE(uint64ParamChoice1 < uint64ParamChoice2);
    ASSERT_TRUE(uint64ParamChoice2 < uint64ParamChoice1);

    uint64ParamChoice2.setA(value);
    ASSERT_FALSE(uint64ParamChoice1 < uint64ParamChoice2);
    ASSERT_FALSE(uint64ParamChoice2 < uint64ParamChoice1);

    const int8_t diffValue = value + 1;
    uint64ParamChoice2.setA(diffValue);
    ASSERT_TRUE(uint64ParamChoice1 < uint64ParamChoice2);
    ASSERT_FALSE(uint64ParamChoice2 < uint64ParamChoice1);
}

TEST_F(UInt64ParamChoiceTest, hashCode)
{
    UInt64ParamChoice uint64ParamChoice1;
    uint64ParamChoice1.initialize(VARIANT_A_SELECTOR);
    UInt64ParamChoice uint64ParamChoice2;
    uint64ParamChoice2.initialize(VARIANT_A_SELECTOR);
    ASSERT_EQ(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

    const int8_t value = 99;
    uint64ParamChoice1.setA(value);
    ASSERT_NE(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

    uint64ParamChoice2.setA(value);
    ASSERT_EQ(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

    const int8_t diffValue = value + 1;
    uint64ParamChoice2.setA(diffValue);
    ASSERT_NE(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

    // use hardcoded values to check that the hash code is stable
    ASSERT_EQ(31623, uint64ParamChoice1.hashCode());
    ASSERT_EQ(31624, uint64ParamChoice2.hashCode());
}

TEST_F(UInt64ParamChoiceTest, writeRead)
{
    uint64_t selector = VARIANT_A_SELECTOR;
    UInt64ParamChoice uint64ParamChoiceA;
    uint64ParamChoiceA.initialize(selector);
    const int8_t valueA = 99;
    uint64ParamChoiceA.setA(valueA);
    zserio::BitStreamWriter writerA(bitBuffer);
    uint64ParamChoiceA.write(writerA);

    zserio::BitStreamReader readerA(writerA.getWriteBuffer(), writerA.getBitPosition(), zserio::BitsTag());
    UInt64ParamChoice readUInt64ParamChoiceA(readerA, selector);
    ASSERT_EQ(valueA, readUInt64ParamChoiceA.getA());

    selector = VARIANT_B_SELECTOR;
    UInt64ParamChoice uint64ParamChoiceB;
    uint64ParamChoiceB.initialize(selector);
    const int16_t valueB = 234;
    uint64ParamChoiceB.setB(valueB);
    zserio::BitStreamWriter writerB(bitBuffer);
    uint64ParamChoiceB.write(writerB);

    zserio::BitStreamReader readerB(writerB.getWriteBuffer(), writerB.getBitPosition(), zserio::BitsTag());
    UInt64ParamChoice readUInt64ParamChoiceB(readerB, selector);
    ASSERT_EQ(valueB, readUInt64ParamChoiceB.getB());
}

TEST_F(UInt64ParamChoiceTest, writeReadFile)
{
    uint64_t selector = VARIANT_A_SELECTOR;
    UInt64ParamChoice uint64ParamChoiceA;
    const int8_t valueA = 99;
    uint64ParamChoiceA.setA(valueA);
    const std::string fileNameA = BLOB_NAME_BASE + "a.blob";
    zserio::serializeToFile(uint64ParamChoiceA, fileNameA, selector);

    const UInt64ParamChoice readUInt64ParamChoiceA =
            zserio::deserializeFromFile<UInt64ParamChoice>(fileNameA, selector);
    ASSERT_EQ(valueA, readUInt64ParamChoiceA.getA());

    selector = VARIANT_B_SELECTOR;
    UInt64ParamChoice uint64ParamChoiceB;
    const int16_t valueB = 234;
    uint64ParamChoiceB.setB(valueB);
    const std::string fileNameB = BLOB_NAME_BASE + "b.blob";
    zserio::serializeToFile(uint64ParamChoiceB, fileNameB, selector);

    const UInt64ParamChoice readUInt64ParamChoiceB =
            zserio::deserializeFromFile<UInt64ParamChoice>(fileNameB, selector);
    ASSERT_EQ(valueB, readUInt64ParamChoiceB.getB());
}

} // namespace uint64_param_choice
} // namespace choice_types
